Merging the code of Polymake/ into Polytopes/

lkastner opened the pull request:
Currently I am simplifying the code of the `Polymake/` folder to use more of the `Polytopes` code. Nevertheless I need input on where the code should end up after refactoring. Not all of it seems to fit perfectly into `Polytopes`.

lkastner committed:
Polytopes: remove zero rows of input as convenience for users

lkastner committed:
Polytopes: Add {interior,boundary}_lattice_points

lkastner committed:
Polytopes: First round of simplifying Polymake/ code

fieker commented on src/Polymake/Ineq.jl:
```diff
@@ -31,121 +31,56 @@ function _polytope(; A::fmpz_mat=zero_matrix(FlintZZ, 1, 1), b::fmpz_mat=zero_ma
-    end
-  end
-  return res
+  return _polytope_new(A = A, b = b, C = identity_matrix(FlintZZ, ncols(A)))
```
Are you sure this returns solutions? It looks like the the polytope itself is returned here...

lkastner replied:
This is just a temporary name. I will change it to something better. I used this to replace `_polytope` in some places, but it did not work out in all places yet.

fieker commented on src/Polymake/Ineq.jl:
```diff
@@ -31,121 +31,56 @@ function _polytope(; A::fmpz_mat=zero_matrix(FlintZZ, 1, 1), b::fmpz_mat=zero_ma
-  else
-    zbA = Matrix{BigInt}(undef, 0, 0)
-  end
-  if !iszero(C)
```
are you (PM) able to handle zero rows?

lkastner replied:
Yes, I actually built this in just now for this occasion.

fieker commented:
to exercise some of the code:
```julia
k, a = quadratic_field(10)
zk = maximal_order(k)
factorisations(15*zk(a))
irreducibles(collect(keys(factor(15*zk))))
norm_equation(zk, 6)
```
by changing the `15` and `6` you can make polymakes life harder...

fieker commented:
I have absolutely no opinion about the final resting place of this code snippet. `git grep` and `methods` will find it!

lkastner committed:
Polymake: Use proper Polyhedron and Cone constructors

- Redirect solve_ineq and solve_non_negative via solve_mixed
- Simplify output parsing
- Reduce code duplication

lkastner committed:
Polytopes: This function takes no b

lkastner committed:
Finishing touches for mergine Polymake into Polytopes

- solve_* has been migrated into Polytopes
- Rename Polymake->NumberTheory (seemed to make sense)
- Add examples to solve_*

lkastner committed:
Polytopes: Add solve_* functions to docs

lkastner committed:
Polytopes: We collect all lattice points anyway

lkastner commented:
I moved the refactored `Ineq.jl` into the `Polytopes` folder. As for the `NmbThy.jl`: I renamed the `Polymake` folder to `NumberTheory`. This seemed to make more sense given the name of the file. There is still the following comment:
```
  #solving Ax = b for x >=0  and A >=0 implies that colums of A with
  #entries > those in b can never be part of a solution
  #thus we prune them
  #of course, this could/ should be done in solve_non_negative...
  #hence in polymake
 ```
 I think `solve_non_negative` is doing the right thing now, but it was unclear to me what needs to be replaced in the function containing that comment. Also there was no documentation in `NmbThy.jl`, so I did not want to mess with it too much.

Nevertheless, I added documentation and examples to the `Ineq.jl`. So I think we can merge this now and then further refactor later.

lkastner commented:
I don't know why the docs fail.

thofma commented:
One signature for `solve_mixed` is included twice.

thofma commented:
One signature for `solve_mixed` is included twice.

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
```diff
@@ -420,6 +420,64 @@ function lattice_points(P::Polyhedron)
+        return VectorIterator{PointVector{Polymake.Integer}}(lat_pts[:, 2:end])
+    else
+        throw(ArgumentError("Polyhedron not bounded"))
+    end
```
Normally I would recommend to validate arguments at the start; this leads to less indentation and IMHO makes it easier to follow the code. Of course in trivial cases like this, it's fine either way. So this is a purely optional suggestion
```suggestion
    pm_polytope(P).BOUNDED || throw(ArgumentError("Polyhedron not bounded"))
    lat_pts = pm_polytope(P).INTERIOR_LATTICE_POINTS
    return VectorIterator{PointVector{Polymake.Integer}}(lat_pts[:, 2:end])
```

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
````diff
@@ -420,6 +420,64 @@ function lattice_points(P::Polyhedron)
+```
+"""
+function boundary_lattice_points(P::Polyhedron)
+    if pm_polytope(P).BOUNDED
````
Same remark as above

fingolfin reviewed (approved):
Looks good to me; minor optional remarks added

lkastner commented:
> One signature for `solve_mixed` is included twice.

Thank you. For some reason I did not understand this error message properly.

lkastner committed:
Apply suggestions from code review

Co-authored-by: Max Horn <max@quendi.de>
